samples/dnn/speech_recognition.cpp

```diff
@@ -0,0 +1,581 @@
+            if (it != labels_map.end())
+                hypotheses.push_back(it->second);
+        }
+        return hypotheses;
+    }
+
+};
+
+static string predict(std::vector<std::vector<double>>& features, dnn::Net net, Decoder decoder)
+{
+    // Passes the features through the Jasper model and decodes the output to english transcripts.
+
+    // expand 2d features matrix to 3d Mat
+    std::vector<int> sizes = {1, static_cast<int>(features.size()),
+                              static_cast<int>(features[0].size())};
+    Mat featuresMat(sizes, CV_64F);
+    for(int i = 0; i < featuresMat.size[1]; ++i)
+    {
+        for(int j = 0; j < featuresMat.size[2]; ++j)
+        {
+            featuresMat.at<double>(0, i, j) = features[i][j];
+        }
+    }
+
+    // make prediction
+    net.setInput(featuresMat);
+    Mat output = net.forward();
+
+    // decode output to transcript
+    auto prediction = decoder.decode(output);
+    return prediction;
+}
+
+static int readAudioFile(vector<double>& inputAudio, string file, int audioStream)
+{
+    VideoCapture cap;
+    int samplingRate = 16000;
+    vector<int> params {    CAP_PROP_AUDIO_STREAM, audioStream,
+                            CAP_PROP_VIDEO_STREAM, -1,
+                            CAP_PROP_AUDIO_DATA_DEPTH, CV_32F,
+                            CAP_PROP_AUDIO_SAMPLES_PER_SECOND, samplingRate
+                            };
+    cap.open(file, CAP_ANY, params);
+    if (!cap.isOpened())
+    {
+        cerr << "Error : Can't read audio file: '" << file << "' with audioStream = " << audioStream << endl;
+        return -1;
+    }
+    const int audioBaseIndex = (int)cap.get(CAP_PROP_AUDIO_BASE_INDEX);
+    vector<double> frameVec;
+    Mat frame;
+    for (;;)
+    {
+        if (cap.grab())
+        {
+            cap.retrieve(frame, audioBaseIndex);
+            frameVec = frame;
+            inputAudio.insert(inputAudio.end(), frameVec.begin(), frameVec.end());
+        }
+        else
+        {
+            break;
+        }
+    }
+    return samplingRate;
+}
+
+static int readAudioMicrophone(vector<double>& inputAudio, int microTime)
+{
+    VideoCapture cap;
+    int samplingRate = 16000;
+    vector<int> params {    CAP_PROP_AUDIO_STREAM, 0,
+                            CAP_PROP_VIDEO_STREAM, -1,
+                            CAP_PROP_AUDIO_DATA_DEPTH, CV_32F,
+                            CAP_PROP_AUDIO_SAMPLES_PER_SECOND, samplingRate
+                            };
+    cap.open(0, CAP_ANY, params);
+    if (!cap.isOpened())
+    {
+        cerr << "Error : Can't open microphone" << endl;
+        return -1;
+    }
+
+    const double cvTickFreq = getTickFrequency();
+    int64 sysTimeCurr = getTickCount();
+    int64 sysTimePrev = sysTimeCurr;
+
+    const int audioBaseIndex = (int)cap.get(CAP_PROP_AUDIO_BASE_INDEX);
+    vector<double> frameVec;
+    Mat frame;
+    while ((sysTimeCurr - sysTimePrev) / cvTickFreq < microTime)
```

Comment:
How do this check could guarantee that `inputAudio` contains enough samples and we don't lost something somewhere?

---

Check should be based on `inputAudio` content size and sample rate.

Reply:
We should read until we fill buffer with enough data.
We don't need timeout check here as first capture may take some extra time for initialization and we will not have enough data in buffer for processing.